core/drivers/bcm_hwrng.c

```diff
@@ -44,29 +48,74 @@ static void bcm_hwrng_reset(void)
 			RNG_CTRL_MASK, RNG_CTRL_ENABLE);
 }
 
-uint32_t bcm_hwrng_read_rng(uint32_t *p_out, uint32_t words_to_read)
+static uint32_t do_rng_read(void *buf, ssize_t blen)
 {
-	uint32_t available_words = 0;
-	uint32_t num_words = 0;
-	uint32_t i = 0;
-	uint64_t timeout = timeout_init_us(RNG_TIMEOUT_US);
+	uint32_t available = 0;
+	ssize_t copied = 0;
+	ssize_t copy_len = 0;
+	union {
+		uint32_t word;
+		uint8_t  bytes[4];
+	} data;
+	uint64_t timeout;
 
 	assert(bcm_hwrng_base);
 
-	do {
-		available_words = io_read32(bcm_hwrng_base +
-					    RNG_FIFO_COUNT_OFFSET);
-		available_words = available_words & RNG_FIFO_COUNT_MASK;
-	} while (!available_words && !timeout_elapsed(timeout));
-
-	if ((available_words > 0) && (words_to_read > 0)) {
-		num_words =  MIN(available_words, words_to_read);
-		for (i = 0; i < num_words; i++)
-			p_out[i] = io_read32(bcm_hwrng_base +
-					     RNG_FIFO_DATA_OFFSET);
+	while (copied < blen) {
+		if (!available) {
+			timeout = timeout_init_us(RNG_TIMEOUT_US);
+			do {
+				available = io_read32(bcm_hwrng_base +
```

Comment:
Are concurrent reads OK here or do we need to protect this function with a lock?

Reply:
@bharat-gooty any comment on this? What happens with multiple CPUs trying to read concurrently?

Reply:
Sorry, I missed it. Yes, need to protect with locks for multiple CPUs read concurrently.

Reply:
Done, added spinlock around hw access.